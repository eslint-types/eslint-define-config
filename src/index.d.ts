import type { Environments } from './env';
import type { Extends } from './extends';
import type { Overrides } from './overrides';
import type { Parser, ParserOptions } from './parser-options';
import type { Plugin } from './plugin';
import type { Rules } from './rules';
import type { Settings } from './settings';

/**
 * ESLint Configuration.
 *
 * @see [ESLint Configuration](https://eslint.org/docs/user-guide/configuring/)
 */
export interface EslintConfig {
  /**
   * @see [Using Configuration Files](https://eslint.org/docs/user-guide/configuring/configuration-files#using-configuration-files)
   */
  root?: boolean;
  /**
   * Tell ESLint to ignore specific files and directories.
   *
   * @see [Ignore Patterns](https://eslint.org/docs/user-guide/configuring/ignoring-code)
   */
  ignorePatterns?: string[];
  /**
   * An environment provides predefined global variables.
   *
   * @see [Environments](https://eslint.org/docs/user-guide/configuring/language-options#specifying-environments)
   */
  env?: Environments;
  /**
   * Extending Configuration Files.
   *
   * @see [Extends](https://eslint.org/docs/user-guide/configuring/configuration-files#extending-configuration-files)
   */
  extends?: Extends;
  /**
   * Specifying Globals.
   *
   * @see [Globals](https://eslint.org/docs/user-guide/configuring/language-options#specifying-globals)
   */
  globals?: Record<
    string,
    'readonly' | 'writable' | false | 'readable' | true | 'writeable' | 'off'
  >;
  /**
   * Parser.
   *
   * @see [Working with Custom Parsers](https://eslint.org/docs/developer-guide/working-with-custom-parsers)
   * @see [Specifying Parser](https://eslint.org/docs/user-guide/configuring/plugins#specifying-parser)
   */
  parser?: Parser;
  /**
   * Parser Options.
   *
   * @see [Working with Custom Parsers](https://eslint.org/docs/developer-guide/working-with-custom-parsers)
   * @see [Specifying Parser Options](https://eslint.org/docs/user-guide/configuring/language-options#specifying-parser-options)
   */
  parserOptions?: ParserOptions;
  /**
   * Which third-party plugins define additional rules, environments, configs, etc. for ESLint to use.
   *
   * @see [Configuring Plugins](https://eslint.org/docs/user-guide/configuring/plugins#configuring-plugins)
   */
  plugins?: Plugin[];
  /**
   * Specifying Processor.
   *
   * @see [processor](https://eslint.org/docs/user-guide/configuring/plugins#specifying-processor)
   */
  processor?: string;
  /**
   * Rules.
   *
   * @see [Rules](https://eslint.org/docs/user-guide/configuring/rules)
   */
  rules?: Rules;
  /**
   * Overrides.
   *
   * @see [How do overrides work](https://eslint.org/docs/user-guide/configuring/configuration-files#how-do-overrides-work)
   */
  overrides?: Overrides;
  /**
   * Settings.
   *
   * @see [Settings](https://eslint.org/docs/user-guide/configuring/configuration-files#adding-shared-settings)
   */
  settings?: Settings;
  /**
   * Disabling Inline Comments.
   *
   * @see [Disabling Inline Comments](https://eslint.org/docs/user-guide/configuring/rules#disabling-inline-comments)
   */
  noInlineConfig?: boolean;
  /**
   * Report unused `eslint-disable` comments.
   *
   * @see [Report unused `eslint-disable` comments](https://eslint.org/docs/user-guide/configuring/rules#report-unused-eslint-disable-comments)
   */
  reportUnusedDisableDirectives?: boolean;
}

/**
 * Define an eslint config.
 *
 * @param config Eslint config.
 * @returns Eslint config.
 */
export function defineConfig(config: EslintConfig): EslintConfig;
