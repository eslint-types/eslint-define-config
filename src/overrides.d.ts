import type { Environments } from './env';
import type { Extends } from './extends';
import type { Parser, ParserOptions } from './parser-options';
import type { Plugin } from './plugin';
import type { Rules } from './rules';
import type { Settings } from './settings';

/**
 *
 */
export interface Override {
  files: string[] | string;
  excludedFiles?: string;
  /**
   * An environment provides predefined global variables.
   *
   * @see [Environments](https://eslint.org/docs/user-guide/configuring/language-options#specifying-environments)
   */
  env?: Environments;
  /**
   * Extending Configuration Files.
   *
   * @see [Extends](https://eslint.org/docs/user-guide/configuring/configuration-files#extending-configuration-files)
   */
  extends?: Extends;
  /**
   * Parser.
   *
   * @see [Working with Custom Parsers](https://eslint.org/docs/developer-guide/working-with-custom-parsers)
   * @see [Specifying Parser](https://eslint.org/docs/user-guide/configuring/plugins#specifying-parser)
   */
  parser?: Parser;
  /**
   * Parser Options.
   *
   * @see [Working with Custom Parsers](https://eslint.org/docs/developer-guide/working-with-custom-parsers)
   * @see [Specifying Parser Options](https://eslint.org/docs/user-guide/configuring/language-options#specifying-parser-options)
   */
  parserOptions?: ParserOptions;
  /**
   * Which third-party plugins define additional rules, environments, configs, etc. for ESLint to use.
   *
   * @see [Configuring Plugins](https://eslint.org/docs/user-guide/configuring/plugins#configuring-plugins)
   */
  plugins?: Plugin[];
  /**
   * Specifying Processor.
   *
   * @see [processor](https://eslint.org/docs/user-guide/configuring/plugins#specifying-processor)
   */
  processor?: string;
  /**
   * Rules.
   *
   * @see [Rules](https://eslint.org/docs/user-guide/configuring/rules)
   */
  rules?: Rules;
  /**
   * Settings.
   *
   * @see [Settings](https://eslint.org/docs/user-guide/configuring/configuration-files#adding-shared-settings)
   */
  settings?: Settings;
}

/**
 * Overrides.
 */
export type Overrides = Array<Override>;
