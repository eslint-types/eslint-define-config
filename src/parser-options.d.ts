import type { LiteralUnion } from './utility-types';

// Some types copied from `@types/eslint` `Linter.ParserOptions`

/**
 * Any valid ECMAScript version number or 'latest':
 *
 * - A version: es3, es5, es6, es7, es8, es9, es10, es11, es12, es13, ...
 * - A year: es2015, es2016, es2017, es2018, es2019, es2020, es2021, es2022, ...
 * - 'latest'
 *
 * @see https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionsecmaversion
 */
export type EcmaVersion =
  | 3
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 2015
  | 2016
  | 2017
  | 2018
  | 2019
  | 2020
  | 2021
  | 2022
  | 'latest';

/**
 * Set to "script" (default) or "module" if your code is in ECMAScript modules.
 */
export type SourceType = 'script' | 'module';

/**
 * An object indicating which additional language features you'd like to use.
 */
export interface EcmaFeatures extends Partial<Record<string, boolean>> {
  /**
   * Allow `return` statements in the global scope.
   */
  globalReturn?: boolean;

  /**
   * Enable global [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode) (if `ecmaVersion` is 5 or greater).
   */
  impliedStrict?: boolean;

  /**
   * Enable [JSX](https://facebook.github.io/jsx).
   */
  jsx?: boolean;
}

/** Lib. */
export type Lib = LiteralUnion<
  | 'es5'
  | 'es6'
  | 'es2015'
  | 'es7'
  | 'es2016'
  | 'es2017'
  | 'es2018'
  | 'es2019'
  | 'es2020'
  | 'esnext'
  | 'dom'
  | 'dom.iterable'
  | 'webworker'
  | 'webworker.importscripts'
  | 'webworker.iterable'
  | 'scripthost'
  | 'es2015.core'
  | 'es2015.collection'
  | 'es2015.generator'
  | 'es2015.iterable'
  | 'es2015.promise'
  | 'es2015.proxy'
  | 'es2015.reflect'
  | 'es2015.symbol'
  | 'es2015.symbol.wellknown'
  | 'es2016.array.include'
  | 'es2017.object'
  | 'es2017.sharedmemory'
  | 'es2017.string'
  | 'es2017.intl'
  | 'es2017.typedarrays'
  | 'es2018.asyncgenerator'
  | 'es2018.asynciterable'
  | 'es2018.intl'
  | 'es2018.promise'
  | 'es2018.regexp'
  | 'es2019.array'
  | 'es2019.object'
  | 'es2019.string'
  | 'es2019.symbol'
  | 'es2020.bigint'
  | 'es2020.promise'
  | 'es2020.sharedmemory'
  | 'es2020.string'
  | 'es2020.symbol.wellknown'
  | 'es2020.intl'
  | 'esnext.array'
  | 'esnext.symbol'
  | 'esnext.asynciterable'
  | 'esnext.intl'
  | 'esnext.bigint'
  | 'esnext.string'
  | 'esnext.promise'
  | 'esnext.weakref'
  | 'es2016.full'
  | 'es2017.full'
  | 'es2018.full'
  | 'es2019.full'
  | 'es2020.full'
  | 'esnext.full'
  | 'lib'
>;

/** DebugLevel. */
export type DebugLevel =
  | boolean
  | Array<'eslint' | 'typescript' | 'typescript-eslint'>;

/** Parser. */
export type Parser = LiteralUnion<
  'babel-eslint' | '@typescript-eslint/parser' | 'vue-eslint-parser'
>;

/**
 * Parser options.
 *
 * @see [Specifying Parser Options](https://eslint.org/docs/user-guide/configuring/language-options#specifying-parser-options)
 */
export interface ParserOptions extends Partial<Record<string, unknown>> {
  /**
   * Accepts any valid ECMAScript version number or `'latest'`:
   *
   * - A version: es3, es5, es6, es7, es8, es9, es10, es11, es12, es13, ..., or
   * - A year: es2015, es2016, es2017, es2018, es2019, es2020, es2021, es2022, ..., or
   * - `'latest'`
   *
   * When it's a version or a year, the value must be a number - so do not include the `es` prefix.
   *
   * Specifies the version of ECMAScript syntax you want to use. This is used by the parser to determine how to perform scope analysis, and it affects the default
   *
   * @default 2018
   * @see https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionsecmaversion
   */
  ecmaVersion?: EcmaVersion;

  /**
   * Set to "script" (default) or "module" if your code is in ECMAScript modules.
   */
  sourceType?: SourceType;

  /**
   * An object indicating which additional language features you'd like to use.
   */
  ecmaFeatures?: EcmaFeatures;

  jsxPragma?: string;

  jsxFragmentName?: string | null;

  /**
   * @see [lib](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionslib)
   */
  lib?: Lib[];

  comment?: boolean;

  debugLevel?: DebugLevel;

  errorOnTypeScriptSyntacticAndSemanticIssues?: boolean;

  errorOnUnknownASTType?: boolean;

  /**
   * This option allows you to provide one or more additional file extensions which should be considered in the TypeScript Program compilation.
   *
   * The default extensions are `.ts`, `.tsx`, `.js`, and `.jsx`. Add extensions starting with `.`, followed by the file extension.
   * E.g. for a `.vue` file use `"extraFileExtensions: [".vue"]`.
   *
   * @see [extraFileExtensions](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionsextrafileextensions)
   */
  extraFileExtensions?: string[];

  filePath?: string;

  loc?: boolean;

  /**
   * Parser.
   *
   * @see [Working with Custom Parsers](https://eslint.org/docs/developer-guide/working-with-custom-parsers)
   * @see [Specifying Parser](https://eslint.org/docs/user-guide/configuring/plugins#specifying-parser)
   */
  parser?: Parser;

  /**
   * @see [project](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionsproject)
   */
  project?: string | string[];

  /**
   * @see [projectFolderIgnoreList](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionsprojectfolderignorelist)
   */
  projectFolderIgnoreList?: Array<string | RegExp>;

  range?: boolean;

  tokens?: boolean;

  /**
   * This option allows you to provide the root directory for relative tsconfig paths specified in the `project` option above.
   *
   * @see [tsconfigRootDir](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionstsconfigrootdir)
   */
  tsconfigRootDir?: string;

  useJSXTextNode?: boolean;

  /**
   * This option allows you to toggle the warning that the parser will give you if you use a version of TypeScript which is not explicitly supported.
   *
   * @default true
   *
   * @see [warnOnUnsupportedTypeScriptVersion](https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/parser#parseroptionswarnonunsupportedtypescriptversion)
   */
  warnOnUnsupportedTypeScriptVersion?: boolean;

  /**
   * This option allow you to tell parser to act as if `emitDecoratorMetadata: true` is set in `tsconfig.json`, but without [type-aware linting](https://typescript-eslint.io/docs/linting/type-linting/).
   * In other words, you don't have to specify `parserOptions.project` in this case, making the linting process faster.
   *
   * @default undefined
   *
   * @see [emitDecoratorMetadata](https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/parser#parseroptionsemitdecoratormetadata)
   */
  emitDecoratorMetadata?: boolean;

  /**
   * @see [vueFeatures](https://github.com/vuejs/vue-eslint-parser#parseroptionsvuefeatures)
   */
  vueFeatures?: {
    /**
     * You can use `parserOptions.vueFeatures.filter` property to specify whether to parse the Vue2 filter.
     *
     * If you specify `false`, the parser does not parse `|` as a filter.
     *
     * @see [filter](https://github.com/vuejs/vue-eslint-parser#parseroptionsvuefeaturesfilter)
     */
    filter?: boolean;

    /**
     * You can use `parserOptions.vueFeatures.interpolationAsNonHTML` property to specify whether to parse the interpolation as HTML.
     *
     * If you specify `true`, the parser handles the interpolation as non-HTML (However, you can use HTML escaping in the interpolation).
     *
     * @see [interpolationAsNonHTML](https://github.com/vuejs/vue-eslint-parser#parseroptionsvuefeaturesinterpolationasnonhtml)
     */
    interpolationAsNonHTML?: boolean;
  };

  /**
   * @see [templateTokenizer](https://github.com/rashfael/eslint-plugin-vue-pug#usage)
   */
  templateTokenizer?: {
    pug?: LiteralUnion<'vue-eslint-parser-template-tokenizer-pug'>;
  };
}
