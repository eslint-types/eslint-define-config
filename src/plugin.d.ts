import type { LiteralUnion } from './utility-types';

/** Plugin. */
export type Plugin = LiteralUnion<
  | '@typescript-eslint'
  | 'import'
  | 'inclusive-language'
  | 'jsdoc'
  | 'mdx'
  | 'prettier'
  | 'spellcheck'
  | 'unicorn'
  | 'vue'
>;
