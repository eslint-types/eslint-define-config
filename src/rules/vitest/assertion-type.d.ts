import type { RuleConfig } from '../rule-config';

/**
 * Option.
 */
export type AssertionTypeOption = {
  type: 'chai' | 'jest';
};

/**
 * Options.
 */
export type AssertionTypeOptions = [AssertionTypeOption?];

/**
 * Enforce assertion type.
 *
 * @see [assertion-type](https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/assertion-type.md)
 */
export type AssertionTypeRuleConfig = RuleConfig<AssertionTypeOptions>;

/**
 * Enforce assertion type.
 *
 * @see [assertion-type](https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/assertion-type.md)
 */
export interface AssertionTypeRule {
  /**
   * Enforce assertion type.
   *
   * @see [assertion-type](https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/assertion-type.md)
   */
  'vitest/assertion-type': AssertionTypeRuleConfig;
}
