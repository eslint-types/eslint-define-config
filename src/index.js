'use strict';

exports.__esModule = true;
exports.defineConfig = void 0;
exports.defineFlatConfig = void 0;

exports.defineConfig = (config) => config;
exports.defineFlatConfig = (config) => config;
