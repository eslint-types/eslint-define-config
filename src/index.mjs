export function defineConfig(config) {
  return config;
}

export function defineFlatConfig(config) {
  return config;
}
