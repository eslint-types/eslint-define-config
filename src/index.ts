import type { ESLintConfig } from './config';
import type { FlatESLintConfig } from './flat-config';

/**
 * Define an ESLint config.
 *
 * @param config ESLint config.
 * @returns ESLint config.
 */
export function defineConfig(config: ESLintConfig): ESLintConfig {
  return config;
}

/**
 * Define an item of Flat ESLint config.
 *
 * @see [Configuration Files (New)](https://eslint.org/docs/latest/user-guide/configuring/configuration-files-new)
 *
 * @param config an item of Flat ESLint config.
 * @returns an item of Flat ESLint config.
 */
export function defineFlatConfig(config: FlatESLintConfig): FlatESLintConfig;

/**
 * Define a flat ESLint config.
 *
 * @see [Configuration Files (New)](https://eslint.org/docs/latest/user-guide/configuring/configuration-files-new)
 *
 * @param config Flat ESLint config.
 * @returns Flat ESLint config.
 */
export function defineFlatConfig(
  config: ReadonlyArray<FlatESLintConfig>,
): FlatESLintConfig[];

export function defineFlatConfig(config: unknown): unknown {
  return config;
}

export type * from './config';
export type * from './flat-config';
export type * from './parser-options';
export type * from './rules';
