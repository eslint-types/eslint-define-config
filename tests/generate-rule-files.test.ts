/* eslint-disable @typescript-eslint/no-unsafe-argument */

import { Filesystem } from '@poppinss/dev-utils';
import { join } from 'path';
import { afterEach, describe, expect, it } from 'vitest';
import { generateTypeFromSchema } from '../scripts/generate-rule-files/src/json-schema-to-ts';
import { RuleFile } from '../scripts/generate-rule-files/src/rule-file';

const fs = new Filesystem(join(__dirname, '../tmp'));

afterEach(() => fs.cleanup());

describe('Json schema to ts', () => {
  it('Should remove garbage comments', async () => {
    const result = await generateTypeFromSchema(
      {
        type: 'object',
        properties: {
          test: {
            anyOf: [
              {
                type: 'object',
                patternProperties: { '^.+$': { type: 'array' } },
              },
            ],
          },
        },
      },
      'test',
    );

    expect(result).to.not.includes('This interface was referenced by');
    expect(result).to.not.includes('via the `');
  });
});

describe('Rule File', () => {
  it('Main schema', async () => {
    const rule = {
      meta: {
        docs: {
          description: 'My rule description',
          url: 'http://test.com/my-rule',
        },
        schema: [{ type: 'integer', minimum: 0 }],
      },
    };

    const ruleFile = new RuleFile(
      { name: 'my-plugin', rules: {} },
      join(fs.basePath, 'my-plugin'),
      'my-rule',
      rule as any,
    );

    const content = await ruleFile.generate();
    expect(content).toMatchSnapshot();
  });

  it('Two schemas', async () => {
    const rule = {
      meta: {
        docs: {
          description: 'My rule description',
          url: 'http://test.com/my-rule',
        },
        schema: [
          { type: 'integer', minimum: 0 },
          { type: 'string', minLength: 1 },
        ],
      },
    };

    const ruleFile = new RuleFile(
      { name: 'my-plugin', rules: {} },
      join(fs.basePath, 'my-plugin'),
      'my-rule',
      rule as any,
    );

    const content = await ruleFile.generate();
    expect(content).toMatchSnapshot();
  });

  it('Three schemas', async () => {
    const rule = {
      meta: {
        docs: {
          description: 'My rule description',
          url: 'http://test.com/my-rule',
        },
        schema: [
          { type: 'integer', minimum: 0 },
          { type: 'string', minLength: 1 },
          { type: 'boolean' },
        ],
      },
    };

    const ruleFile = new RuleFile(
      { name: 'my-plugin', rules: {} },
      join(fs.basePath, 'my-plugin'),
      'my-rule',
      rule as any,
    );

    const content = await ruleFile.generate();
    expect(content).toMatchSnapshot();
  });

  it('Object schema', async () => {
    const rule = {
      meta: {
        docs: {
          description: 'My rule description',
          url: 'http://test.com/my-rule',
        },
        schema: {
          type: 'object',
        },
      },
    };

    const ruleFile = new RuleFile(
      { name: 'my-plugin', rules: {} },
      join(fs.basePath, 'my-plugin'),
      'my-rule',
      rule as any,
    );

    const content = await ruleFile.generate();
    expect(content).toMatchSnapshot();
  });
});
