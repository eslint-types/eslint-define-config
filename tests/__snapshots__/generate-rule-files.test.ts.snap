// Vitest Snapshot v1

exports[`Rule File > Main schema 1`] = `
"import type { RuleConfig } from '../rule-config';

/**
 * Option.
 */
export type MyRuleOption = number;

/**
 * Options.
 */
export type MyRuleOptions = [MyRuleOption?];

/**
 * My rule description.
 *
 * @see [my-rule](http://test.com/my-rule)
 */
export type MyRuleRuleConfig = RuleConfig<MyRuleOptions>;

/**
 * My rule description.
 *
 * @see [my-rule](http://test.com/my-rule)
 */
export interface MyRuleRule {
  /**
   * My rule description.
   *
   * @see [my-rule](http://test.com/my-rule)
   */
  'my-plugin/my-rule': MyRuleRuleConfig;
}
"
`;

exports[`Rule File > Object schema 1`] = `
"import type { RuleConfig } from '../rule-config';

/**
 * Option.
 */
export interface MyRuleOption {
  [k: string]: any;
}

/**
 * Options.
 */
export type MyRuleOptions = MyRuleOption;

/**
 * My rule description.
 *
 * @see [my-rule](http://test.com/my-rule)
 */
export type MyRuleRuleConfig = RuleConfig<MyRuleOptions>;

/**
 * My rule description.
 *
 * @see [my-rule](http://test.com/my-rule)
 */
export interface MyRuleRule {
  /**
   * My rule description.
   *
   * @see [my-rule](http://test.com/my-rule)
   */
  'my-plugin/my-rule': MyRuleRuleConfig;
}
"
`;

exports[`Rule File > Three schemas 1`] = `
"import type { RuleConfig } from '../rule-config';

/**
 * Setting.
 */
export type MyRuleSetting = boolean;

/**
 * Config.
 */
export type MyRuleConfig = string;

/**
 * Option.
 */
export type MyRuleOption = number;

/**
 * Options.
 */
export type MyRuleOptions = [MyRuleOption?, MyRuleConfig?, MyRuleSetting?];

/**
 * My rule description.
 *
 * @see [my-rule](http://test.com/my-rule)
 */
export type MyRuleRuleConfig = RuleConfig<MyRuleOptions>;

/**
 * My rule description.
 *
 * @see [my-rule](http://test.com/my-rule)
 */
export interface MyRuleRule {
  /**
   * My rule description.
   *
   * @see [my-rule](http://test.com/my-rule)
   */
  'my-plugin/my-rule': MyRuleRuleConfig;
}
"
`;

exports[`Rule File > Two schemas 1`] = `
"import type { RuleConfig } from '../rule-config';

/**
 * Config.
 */
export type MyRuleConfig = string;

/**
 * Option.
 */
export type MyRuleOption = number;

/**
 * Options.
 */
export type MyRuleOptions = [MyRuleOption?, MyRuleConfig?];

/**
 * My rule description.
 *
 * @see [my-rule](http://test.com/my-rule)
 */
export type MyRuleRuleConfig = RuleConfig<MyRuleOptions>;

/**
 * My rule description.
 *
 * @see [my-rule](http://test.com/my-rule)
 */
export interface MyRuleRule {
  /**
   * My rule description.
   *
   * @see [my-rule](http://test.com/my-rule)
   */
  'my-plugin/my-rule': MyRuleRuleConfig;
}
"
`;
